### DISCUSSION TITLE
Impact of non-temporal stores on memory error detection reliability ?

### DISCUSSION THREAD

**Discussion by debrouxl** [memtest86plus/memtest86plus#79] [General]
For "large enough" blocks of memory, where the exact value of "large enough" depends on the ISA and last level cache size but is therefore generally much lower than 1 GB, non-temporal stores, e.g. SSE2 and later `movnti` which exists in 32-bit and 64-bit versions, and has general-purpose register operands unlike MMX+ `movntq`, are documented to be able to increase bandwidth in `memset` & `memcpy` workloads.
But do they have different properties wrt. memory error detection _reliability_, which remains the prime aim of a piece of software like memtest86+, above testing _speed_ ? I'd expect that non-temporal stores are neither better, nor worse than temporal stores, but maybe there are some differences.

If this kind of reliability relevant to a memory testing tool is not reduced (or made better ??!) by non-temporal stores, it _may_ be worth pursuing the approach in the memory filling phases of select tests, e.g. 1 and 2, which are slow on my 4S Opteron Bulldozers, as indicated in #12 . Otherwise, it's a non-starter.

**Comment by debrouxl** [comment]
FTR, as a data point, non-temporal stores make tests 1 and 2 faster, on the usual PowerEdge R815, 4 x Opteron 6378 (4 x 8C / 16T), 32 x 8 GB DDR3-1600 registered ECC DIMMs.

Test 1:
normal: cores 0 and 1: 0'03" -> 3'20" -> 6'35" ~6'32"
movnti: cores 0 and 1: 0'50" -> 3'41" -> 6'35" ~5'45"

Test 2:
normal: cores 0 and 1: 6'50" -> 8'35" -> 10'07" + 10'07'" -> 11'4?" -> 13'25" ~6'35"
movnti: cores 0 and 1: 6'55" -> 8'17" ->  9'49" +  9'50" ->  11'12" -> 12'44" ~5'49"

2 tests * (64 / 2) cores * 45" = 3450", which would translate to nearly an hour of run time saved per pass. That's not earth-shattering on a test which takes over a day, but not negligible either.

Quick and dirty trivial patch used to make the above test:
```
diff --git a/system/memrw32.h b/system/memrw32.h
index e9316ec..a45b81b 100644
--- a/system/memrw32.h
+++ b/system/memrw32.h
@@ -44,6 +44,20 @@ static inline void write32(const volatile uint32_t *ptr, uint32_t val)
     );
 }
 
+/**
+ * Writes val to the 32-bit memory location pointed to by ptr, using non-temporal hint
+ */
+static inline void write32_nt(const volatile uint32_t *ptr, uint32_t val)
+{
+    __asm__ __volatile__(
+        "data32 movnti %1, %0"
+        :
+        : "m" (*ptr),
+          "r" (val)
+        : "memory"
+    );
+}
+
 /**
  * Writes val to the 32-bit memory location pointed to by ptr. Reads it
  * back (and discards it) to ensure the write is complete.
diff --git a/system/memrw64.h b/system/memrw64.h
index f45d36b..69e972b 100644
--- a/system/memrw64.h
+++ b/system/memrw64.h
@@ -44,6 +44,20 @@ static inline void write64(const volatile uint64_t *ptr, uint64_t val)
     );
 }
 
+/**
+ * Writes val to the 64-bit memory location pointed to by ptr.
+ */
+static inline void write64_nt(const volatile uint64_t *ptr, uint64_t val)
+{
+    __asm__ __volatile__(
+        "movnti %1, %0"
+        :
+        : "m" (*ptr),
+          "r" (val)
+        : "memory"
+    );
+}
+
 /**
  * Writes val to the 64-bit memory location pointed to by ptr. Reads it
  * back (and discards it) to ensure the write is complete.
diff --git a/tests/own_addr.c b/tests/own_addr.c
index b5a9223..6f6a32c 100644
--- a/tests/own_addr.c
+++ b/tests/own_addr.c
@@ -58,8 +58,11 @@ static int pattern_fill(int my_cpu, testword_t offset)
             }
             test_addr[my_cpu] = (uintptr_t)p;
             do {
-                write_word(p, (testword_t)p + offset);
+                write_word_nt(p, (testword_t)p + offset);
             } while (p++ < pe); // test before increment in case pointer overflows
+#ifdef __x86_64__
+            __asm__ __volatile__ ("sfence");
+#endif
             do_tick(my_cpu);
             BAILOUT;
         } while (!at_end && ++pe); // advance pe to next start point
diff --git a/tests/test_helper.h b/tests/test_helper.h
index 25a49f1..4f41411 100644
--- a/tests/test_helper.h
+++ b/tests/test_helper.h
@@ -21,12 +21,14 @@
  */
 #ifdef __x86_64__
 #include "memrw64.h"
-#define read_word   read64
-#define write_word  write64
+#define read_word      read64
+#define write_word     write64
+#define write_word_nt  write64_nt
 #else
 #include "memrw32.h"
-#define read_word   read32
-#define write_word  write32
+#define read_word      read32
+#define write_word     write32
+#define write_word_nt  write32_nt
 #endif
 
 /**
```

**Comment by martinwhitaker** [comment]
My initial reaction is that it should make no difference to the error detection reliability.

We should also consider how useful it is to run tests 1 and 2 individually on every core. They are testing each core can access each memory location, but that is testing the CPU, not the memory. Also IIRC, test 2 is only really needed in the 32-bit version.

**Comment by debrouxl** [comment]
That's also testing the memory controllers and the inter-socket communication, yeah.

Running tests 1 and 2 on every individual thread, or even every individual core (hardware threads for a given core are supposed to share the memory controller, right ?), would be less of an issue if it were at least twice faster. Over 2 minutes to write + read 256 GB of memory isn't great, though (especially with non-temporal stores) it is quite consistent with memtest86+'s memory bandwidth estimation of 4.95-5.0 GB/s, which is itself barely lower than the L3 bandwidth estimated around 5.5-5.6 GB/s by memtest86+.
The Opteron 6378 can theoretically do an order of magnitude better ( https://www.cpu-world.com/CPUs/Bulldozer/AMD-Opteron%206378%20-%20OS6378WKTGGHK.html ), and quad-channel DDR PC3-12800 at 2 DIMMs per channel matches that maximum memory bandwidth (when accessing the memory belonging to the current NUMA node, that is).

It's not just the PowerEdge R815: the RS904A with quad Opteron 6272 (4 x 8C / 16T), 16 x 16 GB DDR3-1600 registered ECC DIMMs (1 DPC) has a slightly lower estimated memory bandwidth figure.

**Comment by x86fr** [comment]
Temporal vs non-temporal stores are mainly related to the OOO engine of the CPU and not the DRAM itself. By using non-temporal stores, you're telling the OOO engine that it can bypass the cache coherency because the data stored will not be used soon after by another core. At some time, you need a full barrier or a `mfence` (not just `sfence`) to ensure the serialization (in both 32 and 64 bits) to DRAM. Here, I suspect a quirk between the cache disable function (which is really unexpected on any modern µarch) and the way a non-temporal mov configure the cache internally. I don't really see how a non-temporal store can work without any cache. If the cache is fully disabled, `movnti` should led to an immediate, temporal store. Or more likely (according to your results), it bypass the cache disable setting. 

Anyway, we need a way to test the efficiency of error detection algorithm in real-world. I'm working with the betatester team to build some reference systems with know-bad memory (due to defective chips or to barely stable overclocking). The goal is to check the following points:

1. The time needed to find the first issue.
2. How many issues are detected in a given period of time (let's say 1 hour)

If needed, I can probably ask the RMA Team to find some known-bad memory module for us to debug.

**Comment by debrouxl** [comment]
Sure. The fence is also needed on x86, I didn't bother checking for SSE2 support in my quick test patch.

Since we seem to agree that non-temporal stores shouldn't worsen error detection reliability, should I send a proper patch ? That is, after:
* adding separate `mov` and `movnt` code paths gated by SSE2 detection, AFAICT only in 32-bit mode, as SSE2 is mandated by the 64-bit specification, and I'm not aware that MMX/SSE/SSE2 can be disabled by the BIOS;
* replacing SSE+ `sfence` by SSE2+ `mfence`. I interpreted the unofficial reformatting of the Intel x86 documentation https://www.felixcloutier.com/x86/movnti as `sfence` being enough, but `movnti` requires SSE2 anyway.

**Comment by x86fr** [comment]
A lower efficiency to detect errors while using `movnti` can potentially come from 2 factors:

- The CPU frontend detects that the code does not need to be written to RAM and optimize it out of a store buffer (the reason why test 1 is not cached). That's especially possible because using a non-temporal store explicitly tell the OOO engine that you want a higher level of optimization by bypassing the cache coherency. Unfortunately, it's almost impossible to know exactly how each µarch will handle this (even more with the CD bit set to 1) without checking the hardware lines with a logic analyzer. A simple Quick & Dirty trick is to use Corsair Xpert Memory (I wrote something about them almost 2 decades ago: http://www.x86-secret.com/articles/ram/ramled/ramled-5.htm).  They're only available as DDR2, but it's often enough to check if the accesses are optimized out.
- Non-temporal instructions do not fetch the corresponding cache line from memory. We don't care for test 1, but for test 2, it adds a communication between the memory and the CPU (that may be find wrong if there is an issue with memory). That said, I don't think this will have a big impact on detection efficiency.

Anyway, Test 1->3 are not really critical (failing here means something is REALLY wrong with the memory subsystem).

Decreasing test time is desirable (1 hour in a 1 day pass IS significant) but as Martin say, without adding complexity and different code paths, setting Test #1 to `ONE` and Test #2 on to 32-bit only will probably be as efficient to reduce test time:

```
--- a/tests/tests.c
+++ b/tests/tests.c
@@ -50,9 +50,13 @@

 test_pattern_t test_list[NUM_TEST_PATTERNS] = {
     // ena,  cpu, stgs, itrs, errs, description
-    { true,  SEQ,    1,    6,    0, "[Address test, walking ones, no cache] "},
-    { true,  SEQ,    1,    6,    0, "[Address test, own address in window]  "},
+    { true,  ONE,    1,    6,    0, "[Address test, walking ones, no cache] "},
+#if TESTWORD_WIDTH > 32
     { true,  SEQ,    2,    6,    0, "[Address test, own address + window]   "},
+#else
+    { true,  SEQ,    1,    6,    0, "[Address test, own address in window]  "},
+#endif
     { true,  PAR,    1,    6,    0, "[Moving inversions, 1s & 0s]           "},
     { true,  PAR,    1,    3,    0, "[Moving inversions, 8 bit pattern]     "},
```

**Comment by debrouxl** [comment]
Having the own address test run on a single core of every memory controller should be enough to test reachability of all memory, but as for complexity, of course... :)

**Comment by x86fr** [comment]
Every core is able to reach every memory controller transparently, even if the memory is physically connected to another CPU on another Socket.

**Comment by debrouxl** [comment]
Yes, it's transparent, albeit with a performance impact described approximately in the ACPI SLIT.

I'll try to put my thoughts another way. CPU-World indicates that the Opteron 6378 has a single, quad-channel memory controller, whereas the Opteron 6272 has two dual-channel memory controllers. Therefore, my 4S server with the former has 4 memory controllers and 4 groups of 8 cores / 16 threads symmetric wrt. local / remote memory controller usage, whereas my 4S servers with the latter have 8 memory controllers and 8 groups of 4 cores / 8 threads.
As such, I'd expect that reachability of all local + remote memory can be adequately tested by only 1 core per socket on Opteron 6378, and 2 suitably chosen cores per socket on Opteron 6272 - thereby making test 1 respectively 8x (no SMT) / 16x (SMT) and 4x (no SMT) / 8x (SMT) faster. It's not enough to test one core per socket, because one socket can have multiple memory controllers.
If it comes to that for some reason, even testing one core per NUMA node (both server models have 8) would still make test 1 4x (no SMT) / 8x (SMT) faster than the current situation.

**Comment by x86fr** [comment]
> I'd expect that reachability of all local + remote memory can be adequately tested by only 1 core per socket 

The reachability of all local + remote memory can be tested by EVERY core from EVERY Socket, not just by 1 core per socket. There is a performance penalty for remote (another socket) memory accesses (mainly on latency, very few on bandwidth) but this penalty is almost unnoticeable for a software like Memtest86+. Well, as long as you only use 1 core to access memory. With many cores trying to access remote memory simultaneously, contention can make things much slower (as already discussed in #70).

From a DRAM point of view, there is no difference between a data coming from a local core and a data coming from a remote core. Testing all the interconnections between NUMA nodes is possible, but it's unrelated to memory (and broken nodes interconnects will probably fail to boot).

**Comment by debrouxl** [comment]
> The reachability of all local + remote memory can be tested by EVERY core from EVERY Socket, not just by 1 core per socket.

Of course, but it doesn't yield better test coverage for memory reachability than just 1 adequately chosen core per socket / 1 core per memory controller, does it ?

> this penalty is almost unnoticeable for a software like Memtest86+

Easily noticeable during tests 1 and 2 on my servers, as indicated in #12 .

**Comment by x86fr** [comment]
> Of course, but it doesn't yield better test coverage for memory reachability than just 1 adequately chosen core per socket / 1 core per memory controller, does it ?

From a memory reachability point of view, all cores are identical. A local core is not better (nor worst) than a remote core.

> Easily noticeable during tests 1 and 2 on my servers, as indicated in https://github.com/memtest86plus/memtest86plus/discussions/12 .

Right, from a performance point of view, it seems that on some configuration, enabling NUMA can sometimes led to abysmall bandwidth between remote and local cores: https://community.intel.com/t5/Software-Tuning-Performance/Memory-bandwidth-on-a-NUMA-system/td-p/1095836. To get the best performance, a NUMA-aware test can be nice, but it's a big task.

**Comment by debrouxl** [comment]
Wow. At least, the NUMA performance difference I see on my servers isn't as bad as the one reported in that thread. I haven't measured the time each of the 8 slices of test 1 / test 2 for a given core takes, but my feeling is that the differences across memory areas are close enough to the data contained in the ACPI SLIT, i.e. that this table contains reasonably accurate information.
